hcp dynamic credentials guide (#620)
hashicorp/terraform-docs-common: website/docs/cloud-docs/workspaces/dynamic-provider-credentials/hcp-configuration.mdx

---
page_title: Dynamic Credentials with the HCP Provider - Workspaces - HCP Terraform
description: >-
  Use OpenID Connect to get short-term credentials for the HCP Terraform provider in
  your HCP Terraform runs.
---

# Dynamic Credentials with the HCP Provider

~> **Important:** If you are self-hosting [Terraform Cloud agents](/terraform/cloud-docs/agents), ensure your agents use [v1.15.1](/terraform/cloud-docs/agents/changelog#1-15-1-05-01-2024) or above. To use the latest dynamic credentials features, [upgrade your agents to the latest version](/terraform/cloud-docs/agents/changelog).

You can use HCP Terraform’s native OpenID Connect integration with HCP to get [dynamic credentials](/terraform/cloud-docs/workspaces/dynamic-provider-credentials) for the HCP provider in your HCP Terraform runs. Configuring the integration requires the following steps:

> 1. **[Configure HCP](#configure-hcp):** Set up a trust configuration between HCP and HCP Terraform. Then, you must create an [service principal in HPC](https://developer.hashicorp.com/hcp/docs/hcp/admin/iam/service-principals) for your HCP Terraform workspaces.

Reviewer: ```suggestion
1. **[Configure HCP](#configure-hcp):** Set up a trust configuration between HCP and HCP Terraform. Then, you must create a [service principal in HPC](https://developer.hashicorp.com/hcp/docs/hcp/admin/iam/service-principals) for your HCP Terraform workspaces.
```

2. **[Configure HCP Terraform](#configure-hcp-terraform):** Add environment variables to the HCP Terraform workspaces where you want to use Dynamic Credentials.

Once you complete the setup, HCP Terraform automatically authenticates to HCP during each run.

## Configure HCP

> You must enable and configure a workload identity pool and provider on HCP. These instructions use the HCP CLI, but you can also use Terraform to configure HCP. Refer to our [example Terraform configuration](https://github.com/hashicorp/terraform-dynamic-credentials-setup-examples/tree/main/hcp).

Author: Sorry about that! We wanted to hide it until we were absolutely certain that this feature could be public. It's now live ✅

> #### Create a service principal

Reviewer 2: Non-Blocking Nitpick: Should this be titlecased to match the other headings in the doc?:
```suggestion
#### Create a Service Principal
```

> You will need to create a service principal for HCP Terraform to assume during runs. Remember the ID of the service principle as you will need it for the next steps.

Reviewer 2: Suggestion: Should be "principal" instead of "principle"?:
```suggestion
You will need to create a service principal for HCP Terraform to assume during runs. Remember the ID of the service principal as you will need it for the next steps.
```

```shell

> hcp iam service-principals create hcp-terraform --project=HCP_PROJECT_ID

Reviewer: Do we need to call out that they need to create this service principal within a certain HCP project? Maybe link to where they can figure out their HCP project's ID?

Author: Yeah, good callout 🤔 I added a bit more detail around the project ID. I'm not sure we need to tell users how to get the project ID in this document though, I went and looked a few other examples of documentation from other products (i.e. [this one](https://developer.hashicorp.com/packer/docs/hcp#hcp_project_id) from Packer) that mentions the HCP project ID and it looks they don't go into too much detail either, which makes me believe HCP Project IDs are a basic enough concept that folks can find them easily enough.

Reviewer: Love it! Thank you for adding the perfect amount of detail here. How would you feel about moving the sentences you wrote about the HCP project ID under the code block? Moving them around in this order:

```
Create a service principal for HCP Terraform to assume during runs by running the following HCP command. Note the ID of the service principal you create because you will need it in the following steps. 

hcp iam service-principals create hcp-terraform --project=HCP_PROJECT_ID

For all remaining steps, replace `HCP_PROJECT_ID` with the ID of the project that contains all the resources and workspaces that you want to manage with this service principal....etc.
```

That way folks get the command _and_ have the direct context for where to use the `HCP_PROJECT_ID`.

```

> You should then grant this service princpal the permissions it needs to manage your infrastructure during runs.

Reviewer: ```suggestion
Grant your service principal the necessary permissions to manage your infrastructure during runs.
```

```shell
hcp projects add-binding \
  --project=HCP_PROJECT_ID \

>   --member=HCP_PRINCIPLE_ID \

Reviewer 2: Suggestion: Should be "principal" instead of "principle"?:
```suggestion
  --member=HCP_PRINCIPAL_ID \
```

  --role=roles/contributor
```

#### Add a Workload Identity Provider

> Now you'll need to create the workload identity provider which HCP will use to authenticate the HCP Terraform run.

Reviewer: ```suggestion
Next, create a workload identity provider that HCP uses to authenticate the HCP Terraform run.
```

```shell
hcp iam workload-identity-providers create-oidc hcp-terraform-dynamic-credentials \
  --service-principal=iam/project/HCP_PROJECT_ID/service-principal/hcp-terraform \
  --issuer=https://app.terraform.io \
  --allowed-audience=hcp.workload.identity
  --conditional-access='jwt_claims.sub matches `^organization:ORG_NAME:project:PROJECT_NAME:workspace:WORKSPACE_NAME:run_phase:.*`' \
  --description="Allow HCP Terraform agents to act as the hcp-terraform service principal"
```

## Configure HCP Terraform

> You’ll need to set some environment variables in your HCP Terraform workspace in order to configure HCP Terraform to authenticate with HCP using dynamic credentials. You can set these as workspace variables, or if you’d like to share one HCP service principal across multiple workspaces, you can use a variable set.

Reviewer: No contractions in docs:
```suggestion
Next, you need to set environment variables in your HCP Terraform workspace to configure HCP Terraform to authenticate with HCP using dynamic credentials. You can set these as workspace variables or use a variable set to share one HCP service principal across multiple workspaces.
```

### Required Environment Variables
| Variable                                    | Value                                                                                                | Notes                                                                                                                                                                                                                                                                                                                     |
|---------------------------------------------|------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `TFC_HCP_PROVIDER_AUTH`                     | `true`                                                                                               | Requires **v1.15.1** or later if self-managing agents. Must be present and set to `true`, or HCP Terraform will not attempt to use dynamic credentials to authenticate to HCP.                                                                                                                                           |

Reviewer: ```suggestion
| `TFC_HCP_PROVIDER_AUTH`                     | `true`                                                                                               | Requires **v1.15.1** or later if you use self-managing agents. Must be present and set to `true`, or HCP Terraform will not attempt to use dynamic credentials to authenticate to HCP.                                                                                                                                           |
```

> | `TFC_HCP_RUN_PROVIDER_RESOURCE_NAME`        | The resource name of the workload identity provider that will be used to assume the service principal| Requires **v1.15.1** or later if self-managing agents. Optional if `PLAN_PROVIDER_RESOURCE_NAME` and `APPLY_PROVIDER_RESOURCE_NAME` are both provided. These variables are described [below](/terraform/cloud-docs/workspaces/dynamic-provider-credentials/hcp-configuration#optional-environment-variables) |

Reviewer: ```suggestion
| `TFC_HCP_RUN_PROVIDER_RESOURCE_NAME`        | The resource name of the workload identity provider that will be used to assume the service principal| Requires **v1.15.1** or later if you use self-managing agents. Optional if you provide `PLAN_PROVIDER_RESOURCE_NAME` and `APPLY_PROVIDER_RESOURCE_NAME`. [Learn more](#optional-environment-variables). |
```

### Optional Environment Variables
You may need to set these variables, depending on your use case.

| Variable                                      | Value                                                                                                                                                                                                                                                                      | Notes                                                                                                                                     |
|-----------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------|

> | `TFC_HCP_WORKLOAD_IDENTITY_AUDIENCE`          | Will be used as the `aud` claim for the identity token. Defaults to the provider resource name for the current run phase that was derived from the values you provided for `RUN_PROVIDER_RESOURCE_NAME`, `PLAN_PROVIDER_RESOURCE_NAME`, and `APPLY_PROVIDER_RESOURCE_NAME` | Requires **v1.15.1** or later if self-managing agents. This is one of the default `aud` formats that HCP accepts.                          |

Reviewer: Active voice:
```suggestion
| `TFC_HCP_WORKLOAD_IDENTITY_AUDIENCE`          | HCP Terraform uses this as the `aud` claim for the identity token. Defaults to the provider resource name for the current run phase, which HCP Terraform derives from the values you provide for `RUN_PROVIDER_RESOURCE_NAME`, `PLAN_PROVIDER_RESOURCE_NAME`, and `APPLY_PROVIDER_RESOURCE_NAME`. | Requires **v1.15.1** or later if you use self-managing agents. This is one of the default `aud` formats that HCP accepts.                          |
```

> | `TFC_HCP_PLAN_PROVIDER_RESOURCE_NAME`         | The resource name of the workload identity provider that will be used to assume the service principal for the plan phase of a run.                                                                                                                                         | Requires **v1.15.1** or later if self-managing agents. Will fall back to the value of `RUN_PROVIDER_RESOURCE_NAME` if not provided.        |

Reviewer: I wasn't quite sure what "will be used to assume" meant here. I tried to make this a bit more clear:
```suggestion
| `TFC_HCP_PLAN_PROVIDER_RESOURCE_NAME`         | The resource name of the workload identity provider that HCP Terraform assumes as the service principal for the plan phase of a run.                                                                                                                                         | Requires **v1.15.1** or later if self-managing agents. If you do not provide a value, defaults to `RUN_PROVIDER_RESOURCE_NAME` value.        |
```

Author: "assumes" as in "assumes the role of the service principal". Good catch though, I'll clarify!

> | `TFC_HCP_APPLY_PROVIDER_RESOURCE_NAME`        | The resource name of the workload identity provider that will be used to assume the service principal for the apply phase of a run.                                                                                                                                        | Requires **v1.15.1** or later if self-managing agents. Will fall back to the value of `RUN_PROVIDER_RESOURCE_NAME` if not provided.        |

Reviewer: ```suggestion
| `TFC_HCP_APPLY_PROVIDER_RESOURCE_NAME`        | The resource name of the workload identity provider that HCP Terraform assumes as the service principal for the apply phase of a run.                                                                                                                                       | Requires **v1.15.1** or later if self-managing agents. If you do not provide a value, defaults to `RUN_PROVIDER_RESOURCE_NAME` value.        |
```

## Configure the HCP Provider

> Make sure that you’re not setting values for the `HCP_CRED_FILE`environment variables as these will conflict with the dynamic credentials authentication process.

Reviewer: Rewrote for active voice, assuming the subject is the HCP provider based on context:
```suggestion
Do not set the `HCP_CRED_FILE` environment variable when configuring the HCP provider, or `HCP_CRED_FILE` will conflict with the dynamic credentials authentication process.
```

### Specifying Multiple Configurations

> ~> **Important:** If you are self-hosting [Terraform Cloud agents](/terraform/cloud-docs/agents), ensure your agents use [v1.15.1](/terraform/cloud-docs/agents/changelog#1-15-1-05-01-2024) or above. To use the latest dynamic credentials features, [upgrade your agents to the latest version](/terraform/cloud-docs/agents/changelog).

Reviewer: ```suggestion
~> **Important:** If you are self-hosting [HCP Terraform agents](/terraform/cloud-docs/agents), ensure your agents use [v1.15.1](/terraform/cloud-docs/agents/changelog#1-15-1-05-01-2024) or above. To use the latest dynamic credentials features, [upgrade your agents to the latest version](/terraform/cloud-docs/agents/changelog).
```

You can add additional configurations to handle multiple distinct HCP setups, enabling you to use multiple [provider aliases](/terraform/language/providers/configuration#alias-multiple-provider-configurations) within the same workspace.

> For more details, see [Specifying Multiple Configurations](/terraform/cloud-docs/workspaces/dynamic-provider-credentials/specifying-multiple-configurations).

Reviewer: "Refer to" is more accessible language for those reading with screen readers:
```suggestion
For more details, refer to [Specifying Multiple Configurations](/terraform/cloud-docs/workspaces/dynamic-provider-credentials/specifying-multiple-configurations).
```

Author: TIL!

#### Required Terraform Variable

> To use additional configurations, add the following code to your Terraform configuration. This lets HCP Terraform supply variable values that you can then use to map authentication and configuration details to the correct provider blocks.

Reviewer: ```suggestion
Add the following variable to your Terraform configuration to set up additional dynamic credential configurations with the HCP provider. This variable lets HCP Terraform supply variable values that you can then use to map authentication and configuration details to the correct provider blocks.
```

```hcl
variable "tfc_hcp_dynamic_credentials" {
  description = "Object containing HCP dynamic credentials configuration"
  type = object({
    default = object({
      credential_file = string
    })
    aliases = map(object({
      credential_file = string
    }))
  })
}
```

#### Example Usage

> 

Reviewer: Do you think it'd be helpful to link the [example Terraform configuration](https://github.com/hashicorp/terraform-dynamic-credentials-setup-examples/tree/main/hcp) down here too?

Author: I think linking it once should be sufficient, and the context for this block here is talking about how to set up dynamic credentials with aliased providers, whereas the example terraform configuration is the most basic/user-friendly example we could create.

Reviewer: Sounds good! Thank you for that extra context!
